### ISSUE TITLE
offlineimap3 crashes with encoding error

### ISSUE THREAD

**Issue by sudipm-mukherjee** [OfflineIMAP/offlineimap3#62]
#### General informations

- system/distribution (with version): Debian
- offlineimap version (`offlineimap -V`): 0.0~git20210225.1e7ef9e+dfsg-3
- server name or domain: riseup.net
- Debian bug: https://bugs.debian.org/986139

#### Logs, error

```
Last 11 debug messages logged for Copy message from Riseup-Remote:origin-mail-dir prior to exception:
thread: Register new thread 'Copy message from Riseup-Remote:origin-mail-dir' (account 'Riseup')
imap:   15:26.95 Copy message from Riseup-Remote:origin-mail-dir [async] UID ('FETCH', '1981', '(BODY.PEEK[])')
imap:   15:26.95 Copy message from Riseup-Remote:origin-mail-dir state_change_pending.acquire
imap:   15:26.95 Copy message from Riseup-Remote:origin-mail-dir state_change_pending.release
imap:   15:26.95 Copy message from Riseup-Remote:origin-mail-dir _request_push(b'EPBL24', UID, {}) = b'EPBL24'
imap:   15:26.95 Copy message from Riseup-Remote:origin-mail-dir data=b'EPBL24 UID FETCH 1981 (BODY.PEEK[])'
imap:   15:26.96 Copy message from Riseup-Remote:origin-mail-dir UID:b'EPBL24'.ready.wait
imap:   15:27.44 Copy message from Riseup-Remote:origin-mail-dir _get_untagged_response(FETCH) => [(b'184 (UID 1981 BODY[] {6664}', b'Return-Path: <bounce-origin-mail-dir@lists.ex
imap:   15:27.44 Copy message from Riseup-Remote:origin-mail-dir _untagged_response(OK, ?, FETCH) => [(b'184 (UID 1981 BODY[] {6664}', b'Return-Path: <bounce-origin-mail-dir@lists.ex
imap: Returned object from fetching 1981: 'Return-Path: <bounce-origin-mail-dir@lists.example.net>
X-Spam-Checker-Version: SpamAssassin 3.4.2 (2018-09-13) on towhee.riseup.net
X-Spam-Level: 
X... v2.0
*****************************************
'
maildir: Write mail 'origin-mail-dir:1981' with flags set()
2021-03-30 10:15:27 INFO: OfflineIMAP 7.3.0
  Licensed under the GNU GPL v2 or any later version (with an OpenSSL exception)
2021-03-30 10:15:27 WARNING: ERROR: Exceptions occurred during the run!
2021-03-30 10:15:27 WARNING: ERROR: Copying message 1981 [acc: Riseup]
  'ascii' codec can't encode characters in position 3523-3525: ordinal not in range(128)
2021-03-30 10:15:27 WARNING: 
Traceback:
  File "/usr/share/offlineimap3/offlineimap/folder/Base.py", line 807, in copymessageto
    new_uid = dstfolder.savemessage(uid, message, flags, rtime)
  File "/usr/share/offlineimap3/offlineimap/folder/Maildir.py", line 380, in savemessage
    tmpname = self.save_to_tmp_file(messagename, msg)
  File "/usr/share/offlineimap3/offlineimap/folder/Maildir.py", line 330, in save_to_tmp_file
    fd.write(msg.as_bytes(policy=output_policy))
  File "/usr/lib/python3.9/email/message.py", line 178, in as_bytes
    g.flatten(self, unixfrom=unixfrom)
  File "/usr/lib/python3.9/email/generator.py", line 116, in flatten
    self._write(msg)
  File "/usr/lib/python3.9/email/generator.py", line 181, in _write
    self._dispatch(msg)
  File "/usr/lib/python3.9/email/generator.py", line 218, in _dispatch
    meth(msg)
  File "/usr/lib/python3.9/email/generator.py", line 268, in _handle_multipart
    self.write(subparts)
  File "/usr/lib/python3.9/email/generator.py", line 410, in write
    self._fp.write(s.encode('ascii', 'surrogateescape'))
```

Please do let me know if you need any additional information and I can ask the Debian user who reported the issue for the details.

**Comment by jishac** [comment]
Are they able to provide a sample or sanitized email that causes the issue?

**Comment by Gu1nness** [comment]
Hi,

Here is a mail that generates the issue 
```
Return-Path: <root@crans.org>
X-Original-To: root@crans.org
Delivered-To: root@crans.org
Received: from redisdead.crans.org (unknown [172.16.10.124])
	(using TLSv1.3 with cipher TLS_AES_256_GCM_SHA384 (256/256 bits)
	 key-exchange X25519 server-signature RSA-PSS (4096 bits))
	(No client certificate requested)
	by zamok.crans.org (Postfix) with ESMTPS id 3953BE013C
	for <root@crans.org>; Sat, 27 Feb 2021 18:56:26 +0100 (CET)
Received: from charybde.crans.org (unknown [IPv6:fd00::10:ea39:35ff:fef0:48c8])
	by redisdead.crans.org (Postfix) with SMTP id 2672C690
	for <root@crans.org>; Sat, 27 Feb 2021 18:56:26 +0100 (CET)
Received: (nullmailer pid 24690 invoked by uid 119);
	Sat, 27 Feb 2021 17:56:26 -0000
Subject: [ftpsync-ubuntu@charybde] ERROR: rsync errors
To: root@crans.org
Auto-Submitted: auto-generated
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary===--27089--16668--16408--==
Date: Sat, 27 Feb 2021 18:56:26 +0100
Message-Id: <1614448586.065316.24678.nullmailer@charybde.crans.org>
From: mirror <mirror@crans.org>

--==--27089--16668--16408--==
Content-type: text/plain; name="rsync-ftpsync-ubuntu.error"

file has vanished: "/dists/xenial-proposed/main/i18n/by-hash/SHA256/5430f9c072d3df36cc3f78798b9c59c927099469651be84ed008618a1a1b9aaa" (in ubuntu)
rsync warning: some files vanished before they could be transferred (code 24) at main.c(1677) [generator=3.1.3]
--==--27089--16668--16408--==
Content-type: text/plain; name="ftpsync-ubuntu.log"

févr. 27 17:55:01 charybde ftpsync-ubuntu[64415]: Mirrorsync start
févr. 27 17:55:01 charybde ftpsync-ubuntu[64415]: Running mirrorsync, update is required, /pool/mirror/ubuntu//Archive-Update-Required-CRANS exists
févr. 27 17:55:01 charybde ftpsync-ubuntu[64415]: Running stage1: rsync --quiet --log-file /var/log/mirror/rsync-ftpsync-ubuntu.log --filter=exclude_/Archive-Update-in-Progress-CRANS --filter=protect_/Archive-Update-in-Progress-CRANS --filter=exclude_/Archive-Update-Required-CRANS --filter=protect_/Archive-Update-Required-CRANS --filter=exclude_/project/trace/CRANS --filter=protect_/project/trace/CRANS --filter=exclude_/project/trace/CRANS-stage1 --filter=protect_/project/trace/CRANS-stage1 --filter=exclude_/project/trace/_hierarchy --filter=protect_/project/trace/_hierarchy --filter=exclude_/project/trace/_traces --filter=protect_/project/trace/_traces --filter=include_/project/ --filter=protect_/project/ --filter=include_/project/trace/ --filter=protect_/project/trace/ --filter=include_/project/trace/* --bwlimit=0 -prltvHSB8192 --safe-links --chmod=D755,F644 --timeout 120 --stats --no-human-readable --no-inc-recursive --exclude=.~tmp~/ archive.ubuntu.com::ubuntu /pool/mirror/ubunt
 u/ --include=*.diff/ --exclude=*.diff/Index --exclude=Packages* --exclude=Sources* --exclude=Release* --exclude=InRelease --include=i18n/by-hash --exclude=i18n/* --exclude=ls-lR*
févr. 27 18:28:40 charybde ftpsync-ubuntu[64415]: Back from rsync with returncode 24
févr. 27 18:28:40 charybde ftpsync-ubuntu[64415]: Running stage2: rsync --quiet --log-file /var/log/mirror/rsync-ftpsync-ubuntu.log --filter=exclude_/Archive-Update-in-Progress-CRANS --filter=protect_/Archive-Update-in-Progress-CRANS --filter=exclude_/Archive-Update-Required-CRANS --filter=protect_/Archive-Update-Required-CRANS --filter=exclude_/project/trace/CRANS --filter=protect_/project/trace/CRANS --filter=exclude_/project/trace/CRANS-stage1 --filter=protect_/project/trace/CRANS-stage1 --filter=exclude_/project/trace/_hierarchy --filter=protect_/project/trace/_hierarchy --filter=exclude_/project/trace/_traces --filter=protect_/project/trace/_traces --filter=include_/project/ --filter=protect_/project/ --filter=include_/project/trace/ --filter=protect_/project/trace/ --filter=include_/project/trace/* --bwlimit=0 -prltvHSB8192 --safe-links --chmod=D755,F644 --timeout 120 --stats --no-human-readable --no-inc-recursive --exclude=.~tmp~/ archive.ubuntu.com::ubuntu /pool/mirror/ubunt
 u/ --max-delete=40000 --delay-updates --delete --delete-delay --delete-excluded
févr. 27 18:55:40 charybde ftpsync-ubuntu[64415]: Back from rsync with returncode 0
févr. 27 18:55:40 charybde ftpsync-ubuntu[64415]: Latest recorded rsync transfer speed: 300 KB/s
févr. 27 18:56:26 charybde ftpsync-ubuntu[64415]: Mirrorsync done
--==--27089--16668--16408--==--
```
with the following stack trace
```
Traceback (most recent call last):
  File "/usr/share/offlineimap3/offlineimap/threadutil.py", line 146, in run
    Thread.run(self)
  File "/usr/lib/python3.9/threading.py", line 892, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/share/offlineimap3/offlineimap/folder/Base.py", line 807, in copymessageto
    new_uid = dstfolder.savemessage(uid, message, flags, rtime)
  File "/usr/share/offlineimap3/offlineimap/folder/Maildir.py", line 380, in savemessage
    tmpname = self.save_to_tmp_file(messagename, msg)
  File "/usr/share/offlineimap3/offlineimap/folder/Maildir.py", line 330, in save_to_tmp_file
    fd.write(msg.as_bytes(policy=output_policy))
  File "/usr/lib/python3.9/email/message.py", line 178, in as_bytes
    g.flatten(self, unixfrom=unixfrom)
  File "/usr/lib/python3.9/email/generator.py", line 116, in flatten
    self._write(msg)
  File "/usr/lib/python3.9/email/generator.py", line 181, in _write
    self._dispatch(msg)
  File "/usr/lib/python3.9/email/generator.py", line 218, in _dispatch
    meth(msg)
  File "/usr/lib/python3.9/email/generator.py", line 268, in _handle_multipart
    self.write(subparts)
  File "/usr/lib/python3.9/email/generator.py", line 410, in write
    self._fp.write(s.encode('ascii', 'surrogateescape'))
UnicodeEncodeError: 'ascii' codec can't encode characters in position 441-442: ordinal not in range(128)
```

**Comment by jishac** [comment]
Hmm, interesting.  Thank you @Gu1nness for the sample email.  I am assuming the email is encoded with 'utf-8' on disk.  With that assumption, I am able to reproduce the exception, both with offlineimap3 and a small test script.  That said, the email has two defects that I can see.
1. The boundary
2. The encoding of the second attachment

It seems the boundary is what is causing the biggest issue for the python email library.  To my understanding, and per the text in [RFC 2046 Section 5.1.1](https://tools.ietf.org/html/rfc2046#section-5.1.1), the boundary in this email should be quoted.  So:

`Content-Type: multipart/mixed; boundary===--27089--16668--16408--==`
should be
`Content-Type: multipart/mixed; boundary="==--27089--16668--16408--=="`

Without the change, the email library will flag a "defect" with the message, and throws the aforementioned UnicodeEncodeError exception when trying to output the parsed message as_bytes (but not as_string interestingly enough).  The defects listed by the python library is `[NoBoundaryInMultipartDefect(), MultipartInvariantViolationDefect()]` which basically means it doesn't see any parts to the message.  With the change to the boundary, the email message parts are parsed correctly and `msg.as_bytes(policy=output_policy)` no longer throws the exception.  

I wonder if the reason is because the library is treating the entire content of the message as "headers" (as it would be if there were no boundary delimiter) which must be ascii, even though the policy is set to 8bit.  That would explain why fixing the boundary will solve the problem as our "8bit" policy applies at that point.

For what it's worth, the second issue is the content encoding for the second attachment
`Content-type: text/plain; name="ftpsync-ubuntu.log"`
should be
`Content-Type: text/plain; charset=utf-8; name="ftpsync-ubuntu.log"`
`Content-Transfer-Encoding: 8bit`
for the content to be correctly interpreted and displayed, again assuming that it actually is encoded as utf-8 on disk.

So how to best approach a fix is interesting.  Should offlineimap attempt to fix messages with broken boundaries, and if so what is the best way?  Does the fact that the method as_string() works when as_bytes() fails in this case of a broken boundary indicate a bug with the python library?  The argument being that it probably should at least give us back what we read in?

**Comment by Gu1nness** [comment]
Hi,

The mail is indeed encoded in UTF-8 : 
```
`---> file 1614448588.3168693_0.zamok:2,
1614448588.3168693_0.zamok:2,: SMTP mail, UTF-8 Unicode text, with very long lines
```

I guess the two following approaches are available:
 * Try to detect the encoding in case of encoding error
 * Use a failover when encoding using `ignore` or `replace` option in the `s.encode` call.

**Comment by jishac** [comment]
@Gu1nness, since the internals now work on bytes instead of strings there should be no need to do string manipulations.  I believe the issue is a bug with the way the Python library behaves in this situation and I have opened a bug there.  However, I have a patch request #63 that should apply a bare minimum fix to the emails such that they will work.  Ideally, the software generating these emails would be fixed to generate proper messages, but the PR will at least handle the situation gracefully.  The idea is that properly quoting the boundary field is not against the spirit of "do not alter the original message".  In fact headers are often manipulated far more in normal transit.

@sudipm-mukherjee you may want to see if this PR fixes the issue for the other users who experienced similar problems.

**Comment by Gu1nness** [comment]
Hi,
I can't find imaplib2=~3.5 which is in requirements.txt so I can't test the patch for now.
It is not available on pypi from what I have seen.

**Comment by sudipm-mukherjee** [comment]
@Gu1nness will it help if I send you a debian package wih the change which you can install and test?

**Comment by Gu1nness** [comment]
Yes indeed!

**Comment by sudipm-mukherjee** [comment]
@Gu1nness mailed you the package now. Sorry for the delay.

**Comment by Gu1nness** [comment]
No problem, I also have my problems.
Unfortunately this still fails with the same email I gave earlier.

**Comment by jishac** [comment]
Hi @Gu1nness , you should see the following type of text in the log:

```
has defects: [NoBoundaryInMultipartDefect(), MultipartInvariantViolationDefect()]
  ... applying multipart boundary fix.
```

I had tested your email specifically with the PR, so I'm curious what the output was (the [updated code](https://github.com/OfflineIMAP/offlineimap3/pull/63) should be much more graceful on errors and will yield the message-id)

**Comment by Gu1nness** [comment]
Well, I have no `has defects` messages in my logs. I'm weirded out.
@sudipm-mukherjee Are you sure you packaged the right branch?

**Comment by sudipm-mukherjee** [comment]
@Gu1nness yes, and I verified again. If you look at the name of the package I sent you then you will find the SHA in it `84b96e1`. And also, to further verify that I have not made any mistake, I extracted the package in a folder and manually checked the files from the commits in `https://github.com/jishac/offlineimap3/commits/encoding_edge_cases`.
You can also extract the package using `dpkg-deb` and `-x` option. Or since you have it installed you can find the files in `/usr/share/offlineimap3` and check if the files there have the correct changes.

**Comment by jishac** [comment]
Hmm, @Gu1nness and @sudipm-mukherjee thank for verifying.  I wonder if I overlooked the direction of the transaction... The patch applies to messages read from the maildir, but I should have put in the same handling for messages coming from an imap server.

It is clear that the internal library want's properly formatted messages and so I think better error handling overall is needed for both maildir and imap handlers for both parsing and output.

I will try to clean up and merge the discussions in this thread and #65 into a more comprehensive error catching solution for these edge cases.  If I apply this boundary correction code to the imap side, I bet it would clean up this bug.

**Comment by jishac** [comment]
@sudipm-mukherjee  and @Gu1nness, I have an updated patch #72 which addresses the boundary fix in both the IMAP and Maildir classes.  I was able to mail myself a similar message with a broken boundary and have tested that the new code handles the  UnicodeEncodeError broken boundary case correctly from either source (previously it would only correct transfers originating from a local maildir).  If you could please review and provide any feedback.  Thanks!

**Comment by sudipm-mukherjee** [comment]
Thanks @jishac. I will prepare the Debian package and send it to @Gu1nness today.

**Comment by sudipm-mukherjee** [comment]
Sorry for the delay. @Gu1nness I have mailed you the deb package now.

**Comment by Gu1nness** [comment]
Testing it right now (sorry, had to grieve a death)

**Comment by sudipm-mukherjee** [comment]
No worries. Sorry to hear about the death.  :(

**Comment by Gu1nness** [comment]
Hi, I'm back. So this patch is working, however I endend up with another crash x)
I'll open an issue for it.
But at least we can eventually package this version.
And I'm patching archvsync which has the encoding errors.

**Comment by thekix** [comment]
Hi @Gu1nness !

Feel free to open the issue directly here in github :-)

Regards,
kix

**Comment by Gu1nness** [comment]
Sure.
I'm pretty sure it could be a good idea to package at least this fixed version of offlineimap3 since bullseye is released tomorrow and it can still be considered an urgent upload.

**Comment by sudipm-mukherjee** [comment]
Thanks for testing @Gu1nness. Bullseye is not possible now, but I will check if I can have it included in the Point Release for 11.1.

**Comment by thekix** [comment]
I am closing this issue because is solved.

Thanks!!
